### PR TITLE
Refactor errors to use @fastify/error and throw correct validation error

### PR DESCRIPTION
**Pull request by Bram-dc**
This pull request changed the plugin to use the internal errors from fastify and also return the correct validation errors.

See official type provider: https://github.com/fastify/fastify-type-provider-typebox/blob/main/index.mts
```
    return {
      // Note: Here we return a FastifySchemaValidationError[] result. As of writing, Fastify
      // does not currently export this type. Future revisions should uncomment the assertion
      // below and return the full set of properties. The specified properties 'message' and
      // 'instancePath' do however result in a near equivalent error messages to Ajv.
      error: errors.map((error) => ({
        message: `${error.message}`,
        instancePath: error.path
      })) // as FastifySchemaValidationError[]
    }
```

- Move ResponseValidationError class to errors.ts
- Add InvalidSchemaError class to errors.ts
- Update imports in core.ts to reflect new file locations
- Update resolveSchema function in core.ts to throw InvalidSchemaError

### CODE REVIEW

File: test/response-schema.spec.ts [hunk 1]
```diff
@@ -47,7 +46,6 @@ describe('response schema', () => {
         return reply.code(500).send({
           error: 'Internal Server Error',
           message: "Response doesn't match the schema",
-          details: (err as unknown as ResponseValidationError).details,
```
Comment (kibertoad) [inline test/response-schema.spec.ts:50]: why can't we return details here and assert that validation error exposes enough details?

Comment (Bram-dc) [reply test/response-schema.spec.ts:50]: There is no details on ResponseValidationError.

We would now have to access it through error.validation

https://fastify.dev/docs/latest/Reference/Validation-and-Serialization/#error-handling
```
fastify.setErrorHandler(function (error, request, reply) {
  if (error.validation) {
     reply.status(422).send(new Error('validation failed'))
  }
})
```

Comment (kibertoad) [reply test/response-schema.spec.ts:50]: that's fine, can we use that field to preserve the initial test logic?

Comment (Bram-dc) [reply test/response-schema.spec.ts:50]: Done, I also renamed ResponseValidationError to ResponseSerializationError since that is what it is.

Comment (Bram-dc) [reply test/response-schema.spec.ts:50]: Validation errors can accessed through error.validation as described in the docs.

https://fastify.dev/docs/latest/Reference/Validation-and-Serialization/#error-handling
```
fastify.setErrorHandler(function (error, request, reply) {
  if (error.validation) {
     reply.status(422).send(new Error('validation failed'))
  }
})
```

File: test/response-schema.spec.ts [hunk 2]
```diff
@@ -71,20 +71,16 @@ describe('response schema', () => {
...
-                "message": "Expected undefined, received object",
-                "path": [],
-                "received": "object",
-              },
-            ],
-            "method": "GET",
-            "url": "/incorrect",
-          },
           "error": "Internal Server Error",
+          "issues": [
```
Comment (kibertoad) [inline test/response-schema.spec.ts:75]: this is less useful than what we had before.

does the new error structure no longer expose the url and the method?

Comment (Bram-dc) [reply test/response-schema.spec.ts:75]: Well we could also throw the old error here, but it is mostly for consistency. This type of error should also result in a 500 Internal server error, so I think it is not important to add the url and method to the error. The default error reporting from fastify reports these values anyways with the error.

Comment (Bram-dc) [reply test/response-schema.spec.ts:75]: image

Plus you can just get those values from the error handler.

Comment (kibertoad) [reply test/response-schema.spec.ts:75]: will need to check it by hand in greater detail to understand whether this is sufficient.

I am primarily concerned about getting good bugsnag entries and logs,  i n c l u d i n g on the client side.

whatever solution we end up with, it should expose no less data than the current one.

File: test/response-schema.spec.ts (hunk 1 above)
Comment (kibertoad) [reply test/response-schema.spec.ts:50]: fastify error model really isn't very flexible :-/
trying to augment errors we are getting there currently

**Review by kibertoad — APPROVED** [review]

**Comment by kibertoad** [comment]
@Bram-dc I've made some adjustments, can you please check if you are OK with them?

File: src/core.ts [hunk 3]
```diff
@@ -118,15 +118,15 @@ export const validatorCompiler: FastifySchemaCompiler<z.ZodTypeAny> =
     return { value: result.data };
   };
 
-const resolveSchema = (maybeSchema: z.ZodTypeAny | { properties: z.ZodTypeAny }) => {
+function resolveSchema(maybeSchema: z.ZodTypeAny | { properties: z.ZodTypeAny }): ZodSchema {
```
Comment (Bram-dc) [inline src/core.ts:121]: Why is this not an arrow function? All others are using an arrow function.

Comment (kibertoad) [reply src/core.ts:121]: only the ones that are exported

File: src/errors.ts [hunk 4]
```diff
@@ -8,12 +8,22 @@ export const ResponseSerializationError = createError<[{ cause: Error }]>(
...
+  };
+};
+
 export const InvalidSchemaError = createError<[string]>(
   'FST_ERR_INVALID_SCHEMA',
   'Invalid schema passed: %s',
   500,
 );
 
+export type InvalidSchemaErrorType = FastifySchemaValidationError;
```
Comment (Bram-dc) [inline src/errors.ts:25]: `FastifySchemaValidationError` is a different error and should not be used here.

```
export interface FastifySchemaValidationError {
  keyword: string;
  instancePath: string;
  schemaPath: string;
  params: Record<string, unknown>;
  message?: string;
}
```

File: src/errors.ts [hunk 5]
```diff
@@ -8,12 +8,22 @@ export const ResponseSerializationError = createError<[{ cause: Error }]>(
   500,
 );
 
+export type ResponseSerializationErrorType = FastifySchemaValidationError & {
```
Comment (Bram-dc) [inline src/errors.ts:11]: Neither here

**Comment by Bram-dc** [comment]
Can I force push an alternative?

**Comment by kibertoad** [comment]
@Bram-dc sure! as long as it preserves the expectations of the test

**Review by kibertoad — APPROVED** [review]

**Comment by kibertoad** [comment]
this is perfect, thank you!

**Pull request merged**